// mezon-sdk/dist/esm/mezon-client/structures/InteractiveMessage.js
import { EMessageComponentType, } from "../../interfaces";
import { getRandomColor } from "../utils/helper";
export class InteractiveBuilder {
    interactive;
    constructor(title) {
        this.interactive = {
            color: getRandomColor(),
            title,
            fields: [],
            timestamp: new Date().toISOString(),
            footer: {
                text: "Powered by Mezon",
                icon_url: "https://cdn.mezon.vn/1837043892743049216/1840654271217930240/1827994776956309500/857_0246x0w.webp",
            },
        };
    }
    setAuthor(name, icon_url, url) {
        this.interactive.author = { name, icon_url, url };
        return this;
    }
    setDescription(description) {
        this.interactive.description = description;
        return this;
    }
    setThumbnail(url) {
        this.interactive.thumbnail = { url };
        return this;
    }
    setImage(url, width, height) {
        this.interactive.image = {
            url,
            width: width ?? "auto",
            height: height ?? "auto",
        };
        return this;
    }
    addField(name, value, inline = false) {
        this.interactive.fields?.push({ name, value, inline });
        return this;
    }
    addInputField(id, name, placeholder, options, description) {
        this.interactive.fields?.push({
            name,
            value: description ?? "",
            inputs: {
                id,
                type: EMessageComponentType.INPUT,
                component: {
                    id: `${id}-component`,
                    placeholder,
                    defaultValue: options?.defaultValue ?? "",
                    type: options?.type ?? "text",
                    textarea: options?.textarea ?? false,
                },
            },
        });
        return this;
    }
    addSelectField(id, name, options, valueSelected, description) {
        this.interactive.fields?.push({
            name,
            value: description ?? "",
            inputs: {
                id,
                type: EMessageComponentType.SELECT,
                component: {
                    options,
                    valueSelected,
                },
            },
        });
        return this;
    }
    addRadioField(id, name, options, description, max_options // Apply when use mutiple choice (number of options)
    ) {
        this.interactive.fields?.push({
            name,
            value: description ?? "",
            inputs: {
                id,
                type: EMessageComponentType.RADIO,
                component: options,
                ...(max_options ? { max_options } : {}),
            },
        });
        return this;
    }
    addDatePickerField(id, name, description) {
        this.interactive.fields?.push({
            name,
            value: description ?? "",
            inputs: {
                id,
                type: EMessageComponentType.DATEPICKER,
                component: {},
            },
        });
        return this;
    }
    addAnimation(id, config, name, description) {
        this.interactive.fields?.push({
            name: name ?? "",
            value: description ?? "",
            inputs: {
                id,
                type: EMessageComponentType.ANIMATION,
                component: config,
            },
        });
        return this;
    }
    build() {
        return this.interactive;
    }
}
