// mezon-sdk/dist/esm/mezon-client/structures/User.js
import { ChannelStreamMode } from "../../constants";
export class User {
    id;
    username;
    clan_nick;
    clan_avatar;
    display_name;
    avartar;
    dmChannelId;
    clan;
    channelManager;
    messageQueue;
    socketManager;
    constructor(initUserData, clan, messageQueue, socketManager, channelManager) {
        this.id = initUserData.id;
        this.avartar = initUserData.avartar ?? "";
        this.dmChannelId = initUserData?.dmChannelId ?? "";
        this.username = initUserData?.username ?? "";
        this.clan_nick = initUserData?.clan_nick ?? "";
        this.clan_avatar = initUserData?.clan_avatar ?? "";
        this.display_name = initUserData?.display_name ?? "";
        this.clan = clan;
        this.channelManager = channelManager;
        this.messageQueue = messageQueue;
        this.socketManager = socketManager;
    }
    async getEphemeralKeyPair() {
        if (!this.clan.mmnClient) {
            throw new Error("MmnClient not initialized");
        }
        return this.clan.mmnClient.generateEphemeralKeyPair();
    }
    async getAddress(user_id) {
        if (!this.clan.mmnClient) {
            throw new Error("MmnClient not initialized");
        }
        return this.clan.mmnClient.getAddressFromUserId(user_id);
    }
    async getZkProofs(data) {
        if (!this.clan.zkClient) {
            throw new Error("ZkClient not initialized");
        }
        const req = {
            userId: data.user_id,
            jwt: data.jwt,
            address: data.address,
            ephemeralPublicKey: data.ephemeral_public_key,
        };
        return this.clan.zkClient.getZkProofs(req);
    }
    async getCurrentNonce(user_id, tag) {
        if (!this.clan.mmnClient) {
            throw new Error("MmnClient not initialized");
        }
        return this.clan.mmnClient.getCurrentNonce(user_id, tag || "pending");
    }
    async sendToken(tokenEvent) {
        if (!this.clan.mmnClient) {
            throw new Error("MmnClient not initialized");
        }
        try {
            return this.clan.mmnClient.sendTransaction({
                sender: tokenEvent.sender_id,
                recipient: tokenEvent.receiver_id,
                amount: this.clan.mmnClient.scaleAmountToDecimals(tokenEvent.amount),
                nonce: tokenEvent.nonce,
                textData: tokenEvent.note,
                extraInfo: tokenEvent.extra_attribute,
                publicKey: tokenEvent.public_key,
                privateKey: tokenEvent.private_key,
                zkProof: tokenEvent.zk_proof,
                zkPub: tokenEvent.zk_pub,
            });
        }
        catch (error) {
            console.log("Error sendToken");
            return null;
        }
    }
    async sendDM(content, code) {
        return this.messageQueue.enqueue(async () => {
            if (!this.dmChannelId) {
                const dmChannel = await this.createDmChannel();
                this.dmChannelId = dmChannel?.channel_id ?? "";
            }
            if (!this.dmChannelId)
                throw Error(`Can not get dmChannelId for this user ${this.id}!`);
            const dataSendDm = {
                clan_id: "0",
                channel_id: this.dmChannelId,
                mode: ChannelStreamMode.STREAM_MODE_DM,
                is_public: false,
                content,
                code,
            };
            return await this.socketManager.writeChatMessage(dataSendDm);
        });
    }
    async createDmChannel() {
        try {
            const dmChannel = await this.channelManager?.createDMchannel(this.id);
            return dmChannel ?? {};
        }
        catch (error) {
            console.log("Error createDmChannel User");
            return null;
        }
    }
    async listTransactionDetail(transactionId) {
        const session = this.clan.sessionToken;
        return this.clan.apiClient.listTransactionDetail(session, transactionId);
    }
}
