// mezon-sdk/dist/cjs/mezon-client/manager/session_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SessionManager = void 0;
const session_1 = require("../../session");
class SessionManager {
    constructor(apiClient, session) {
        this.apiClient = apiClient;
        this.session = session;
    }
    async authenticate(apiKey) {
        return this.apiClient
            .mezonAuthenticate(apiKey, "", {
            account: {
                token: apiKey,
            },
        })
            .then(async (apiSession) => {
            this.session = new session_1.Session(apiSession);
            return this.session;
        });
    }
    async logout() {
        if (!this.session)
            return false;
        const request = {
            token: this.session.token,
            refresh_token: this.session.refresh_token,
        };
        return this.apiClient
            .mezonAuthenticateLogout(this.session.token, request)
            .then((response) => {
            return response !== undefined;
        });
    }
    getSession() {
        return this.session;
    }
}
exports.SessionManager = SessionManager;
