// mezon-sdk/dist/cjs/interfaces/client.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EMarkdownType = exports.EMessageSelectType = exports.EMessageComponentType = exports.EButtonMessageStyle = void 0;
var EButtonMessageStyle;
(function (EButtonMessageStyle) {
    EButtonMessageStyle[EButtonMessageStyle["PRIMARY"] = 1] = "PRIMARY";
    EButtonMessageStyle[EButtonMessageStyle["SECONDARY"] = 2] = "SECONDARY";
    EButtonMessageStyle[EButtonMessageStyle["SUCCESS"] = 3] = "SUCCESS";
    EButtonMessageStyle[EButtonMessageStyle["DANGER"] = 4] = "DANGER";
    EButtonMessageStyle[EButtonMessageStyle["LINK"] = 5] = "LINK";
})(EButtonMessageStyle = exports.EButtonMessageStyle || (exports.EButtonMessageStyle = {}));
var EMessageComponentType;
(function (EMessageComponentType) {
    EMessageComponentType[EMessageComponentType["BUTTON"] = 1] = "BUTTON";
    EMessageComponentType[EMessageComponentType["SELECT"] = 2] = "SELECT";
    EMessageComponentType[EMessageComponentType["INPUT"] = 3] = "INPUT";
    EMessageComponentType[EMessageComponentType["DATEPICKER"] = 4] = "DATEPICKER";
    EMessageComponentType[EMessageComponentType["RADIO"] = 5] = "RADIO";
    EMessageComponentType[EMessageComponentType["ANIMATION"] = 6] = "ANIMATION";
    EMessageComponentType[EMessageComponentType["GRID"] = 7] = "GRID";
})(EMessageComponentType = exports.EMessageComponentType || (exports.EMessageComponentType = {}));
var EMessageSelectType;
(function (EMessageSelectType) {
    EMessageSelectType[EMessageSelectType["TEXT"] = 1] = "TEXT";
    EMessageSelectType[EMessageSelectType["USER"] = 2] = "USER";
    EMessageSelectType[EMessageSelectType["ROLE"] = 3] = "ROLE";
    EMessageSelectType[EMessageSelectType["CHANNEL"] = 4] = "CHANNEL";
})(EMessageSelectType = exports.EMessageSelectType || (exports.EMessageSelectType = {}));
var EMarkdownType;
(function (EMarkdownType) {
    EMarkdownType["TRIPLE"] = "t";
    EMarkdownType["SINGLE"] = "s";
    EMarkdownType["PRE"] = "pre";
    EMarkdownType["CODE"] = "c";
    EMarkdownType["BOLD"] = "b";
    EMarkdownType["LINK"] = "lk";
    EMarkdownType["VOICE_LINK"] = "vk";
    EMarkdownType["LINKYOUTUBE"] = "lk_yt";
})(EMarkdownType = exports.EMarkdownType || (exports.EMarkdownType = {}));
