// mezon-sdk/dist/cjs/mezon-client/structures/InteractiveMessage.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.InteractiveBuilder = void 0;
const interfaces_1 = require("../../interfaces");
const helper_1 = require("../utils/helper");
class InteractiveBuilder {
    constructor(title) {
        this.interactive = {
            color: (0, helper_1.getRandomColor)(),
            title,
            fields: [],
            timestamp: new Date().toISOString(),
            footer: {
                text: "Powered by Mezon",
                icon_url: "https://cdn.mezon.vn/1837043892743049216/1840654271217930240/1827994776956309500/857_0246x0w.webp",
            },
        };
    }
    setAuthor(name, icon_url, url) {
        this.interactive.author = { name, icon_url, url };
        return this;
    }
    setDescription(description) {
        this.interactive.description = description;
        return this;
    }
    setThumbnail(url) {
        this.interactive.thumbnail = { url };
        return this;
    }
    setImage(url, width, height) {
        this.interactive.image = {
            url,
            width: width !== null && width !== void 0 ? width : "auto",
            height: height !== null && height !== void 0 ? height : "auto",
        };
        return this;
    }
    addField(name, value, inline = false) {
        var _a;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({ name, value, inline });
        return this;
    }
    addInputField(id, name, placeholder, options, description) {
        var _a, _b, _c, _d;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({
            name,
            value: description !== null && description !== void 0 ? description : "",
            inputs: {
                id,
                type: interfaces_1.EMessageComponentType.INPUT,
                component: {
                    id: `${id}-component`,
                    placeholder,
                    defaultValue: (_b = options === null || options === void 0 ? void 0 : options.defaultValue) !== null && _b !== void 0 ? _b : "",
                    type: (_c = options === null || options === void 0 ? void 0 : options.type) !== null && _c !== void 0 ? _c : "text",
                    textarea: (_d = options === null || options === void 0 ? void 0 : options.textarea) !== null && _d !== void 0 ? _d : false,
                },
            },
        });
        return this;
    }
    addSelectField(id, name, options, valueSelected, description) {
        var _a;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({
            name,
            value: description !== null && description !== void 0 ? description : "",
            inputs: {
                id,
                type: interfaces_1.EMessageComponentType.SELECT,
                component: {
                    options,
                    valueSelected,
                },
            },
        });
        return this;
    }
    addRadioField(id, name, options, description, max_options // Apply when use mutiple choice (number of options)
    ) {
        var _a;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({
            name,
            value: description !== null && description !== void 0 ? description : "",
            inputs: Object.assign({ id, type: interfaces_1.EMessageComponentType.RADIO, component: options }, (max_options ? { max_options } : {})),
        });
        return this;
    }
    addDatePickerField(id, name, description) {
        var _a;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({
            name,
            value: description !== null && description !== void 0 ? description : "",
            inputs: {
                id,
                type: interfaces_1.EMessageComponentType.DATEPICKER,
                component: {},
            },
        });
        return this;
    }
    addAnimation(id, config, name, description) {
        var _a;
        (_a = this.interactive.fields) === null || _a === void 0 ? void 0 : _a.push({
            name: name !== null && name !== void 0 ? name : "",
            value: description !== null && description !== void 0 ? description : "",
            inputs: {
                id,
                type: interfaces_1.EMessageComponentType.ANIMATION,
                component: config,
            },
        });
        return this;
    }
    build() {
        return this.interactive;
    }
}
exports.InteractiveBuilder = InteractiveBuilder;
