// mezon-sdk/dist/cjs/web_socket_adapter_pb.js
"use strict";
/**
 * Copyright 2020 The Mezon Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __setModuleDefault = (this && this.__setModuleDefault) || (Object.create ? (function(o, v) {
    Object.defineProperty(o, "default", { enumerable: true, value: v });
}) : function(o, v) {
    o["default"] = v;
});
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (k !== "default" && Object.prototype.hasOwnProperty.call(mod, k)) __createBinding(result, mod, k);
    __setModuleDefault(result, mod);
    return result;
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.WebSocketAdapterPb = void 0;
const tsproto = __importStar(require("./rtapi/realtime"));
const ws_1 = __importDefault(require("ws"));
/**
 * A protocol buffer socket adapter that accepts and transmits payloads using the protobuf binary wire format.
 */
class WebSocketAdapterPb {
    constructor() {
    }
    get onClose() {
        return this._socket.onclose;
    }
    set onClose(value) {
        this._socket.onclose = value;
    }
    get onError() {
        return this._socket.onerror;
    }
    set onError(value) {
        this._socket.onerror = value;
    }
    get onMessage() {
        return this._socket.onmessage;
    }
    set onMessage(value) {
        try {
            if (value) {
                this._socket.onmessage = (evt) => {
                    try {
                        const buffer = evt.data;
                        const uintBuffer = new Uint8Array(buffer);
                        const envelope = tsproto.Envelope.decode(uintBuffer);
                        if (envelope.channel_message) {
                            if (envelope.channel_message.code == undefined) {
                                //protobuf plugin does not default-initialize missing Int32Value fields
                                envelope.channel_message.code = 0;
                            }
                        }
                        value(envelope);
                    }
                    catch (e) {
                        console.log(e);
                    }
                };
            }
            else {
                value = null;
            }
        }
        catch (e) {
            console.log(e);
        }
    }
    get onOpen() {
        return this._socket.onopen;
    }
    set onOpen(value) {
        this._socket.onopen = value;
    }
    isOpen() {
        var _a;
        return ((_a = this._socket) === null || _a === void 0 ? void 0 : _a.readyState) == ws_1.default.OPEN;
    }
    close() {
        var _a;
        (_a = this._socket) === null || _a === void 0 ? void 0 : _a.close();
        this._socket = undefined;
    }
    connect(scheme, host, port, createStatus, token, signal) {
        if (signal) {
            signal.addEventListener('abort', () => {
                this.close();
            });
        }
        const url = `${scheme}${host}:${port}/ws?lang=en&status=${encodeURIComponent(createStatus.toString())}&token=${encodeURIComponent(token)}&format=protobuf`;
        this._socket = new ws_1.default(url);
        this._socket.binaryType = "arraybuffer";
    }
    send(msg) {
        if (msg.match_data_send) {
            let payload = msg.match_data_send.data;
            // can't send a string over protobuf
            if (typeof payload == "string") {
                msg.match_data_send.data = new TextEncoder().encode(payload);
            }
        }
        else if (msg.party_data_send) {
            let payload = msg.party_data_send.data;
            // can't send a string over protobuf
            if (typeof payload == "string") {
                msg.party_data_send.data = new TextEncoder().encode(payload);
            }
        }
        const envelopeWriter = tsproto.Envelope.encode(tsproto.Envelope.fromPartial(msg));
        const encodedMsg = envelopeWriter.finish();
        this._socket.send(encodedMsg);
    }
}
exports.WebSocketAdapterPb = WebSocketAdapterPb;
