// mezon-sdk/dist/cjs/mezon-client-v2/utils/CacheManager.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.CacheManager = void 0;
const Collection_1 = __importDefault(require("./Collection"));
class CacheManager {
    constructor(fetcher, maxSize = Infinity) {
        this.fetcher = fetcher;
        this.maxSize = maxSize;
        this.cache = new Collection_1.default();
    }
    get size() {
        return this.cache.size;
    }
    get(id) {
        return this.cache.get(id);
    }
    set(id, value) {
        if (this.cache.size >= this.maxSize) {
            const firstKey = this.cache.firstKey();
            if (firstKey)
                this.cache.delete(firstKey);
        }
        this.cache.set(id, value);
    }
    async fetch(id) {
        const existing = this.get(id);
        if (existing)
            return existing;
        const fetched = await this.fetcher(id);
        this.set(id, fetched);
        return fetched;
    }
    first() {
        return this.cache.first();
    }
    filter(fn) {
        return this.cache.filter(fn);
    }
    map(fn) {
        return this.cache.map(fn);
    }
    values() {
        return this.cache.values();
    }
    delete(id) {
        return this.cache.delete(id);
    }
}
exports.CacheManager = CacheManager;
