// mezon-sdk/dist/cjs/mezon-client-v2/structures/User.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.User = void 0;
class User {
    constructor(initUserData, clan, channelManager) {
        this.id = initUserData.id;
        this.avartar = initUserData.avartar;
        this.dmChannelId = initUserData === null || initUserData === void 0 ? void 0 : initUserData.dmChannelId;
        this.username = initUserData === null || initUserData === void 0 ? void 0 : initUserData.username;
        this.clan_nick = initUserData === null || initUserData === void 0 ? void 0 : initUserData.clan_nick;
        this.clan_avatar = initUserData === null || initUserData === void 0 ? void 0 : initUserData.clan_avatar;
        this.display_name = initUserData === null || initUserData === void 0 ? void 0 : initUserData.display_name;
        this.clan = clan;
        this.channelManager = channelManager;
    }
    async sendToken(sendTokenData) {
        var _a, _b;
        const dataSendToken = {
            receiver_id: this.id,
            amount: sendTokenData.amount,
            note: (_a = sendTokenData === null || sendTokenData === void 0 ? void 0 : sendTokenData.note) !== null && _a !== void 0 ? _a : "",
            extra_attribute: (_b = sendTokenData === null || sendTokenData === void 0 ? void 0 : sendTokenData.extra_attribute) !== null && _b !== void 0 ? _b : "",
        };
        return this.clan.apiClient.sendToken(this.clan.sessionToken, dataSendToken);
    }
    async _createDmChannel() {
        const dmChannel = await this.channelManager.createDMchannel(this.id);
        return dmChannel !== null && dmChannel !== void 0 ? dmChannel : {};
    }
}
exports.User = User;
