// mezon-sdk/dist/cjs/mezon-client/event_manager.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.EventManager = void 0;
class EventManager {
    constructor() {
        this.events = {};
    }
    on(event, func, context = null) {
        if (!this.events[event]) {
            this.events[event] = [];
        }
        if (typeof func !== "function") {
            throw new Error("Please add a function to the event");
        }
        this.events[event].push(context ? func.bind(context) : func);
    }
    remove(event, func) {
        if (!this.events[event]) {
            throw new Error("Event not supported");
        }
        this.events[event] = this.events[event].filter((f) => f !== func);
    }
    emit(event, ...args) {
        if (this.events[event]) {
            this.events[event].forEach((func) => {
                try {
                    func(...args);
                }
                catch (err) {
                    console.error(`Error executing event handler for ${event}:`, err);
                }
            });
        }
    }
}
exports.EventManager = EventManager;
