// mezon-sdk/dist/esm/interfaces/client.js
export var EButtonMessageStyle;
(function (EButtonMessageStyle) {
    EButtonMessageStyle[EButtonMessageStyle["PRIMARY"] = 1] = "PRIMARY";
    EButtonMessageStyle[EButtonMessageStyle["SECONDARY"] = 2] = "SECONDARY";
    EButtonMessageStyle[EButtonMessageStyle["SUCCESS"] = 3] = "SUCCESS";
    EButtonMessageStyle[EButtonMessageStyle["DANGER"] = 4] = "DANGER";
    EButtonMessageStyle[EButtonMessageStyle["LINK"] = 5] = "LINK";
})(EButtonMessageStyle || (EButtonMessageStyle = {}));
export var EMessageComponentType;
(function (EMessageComponentType) {
    EMessageComponentType[EMessageComponentType["BUTTON"] = 1] = "BUTTON";
    EMessageComponentType[EMessageComponentType["SELECT"] = 2] = "SELECT";
    EMessageComponentType[EMessageComponentType["INPUT"] = 3] = "INPUT";
    EMessageComponentType[EMessageComponentType["DATEPICKER"] = 4] = "DATEPICKER";
    EMessageComponentType[EMessageComponentType["RADIO"] = 5] = "RADIO";
    EMessageComponentType[EMessageComponentType["ANIMATION"] = 6] = "ANIMATION";
    EMessageComponentType[EMessageComponentType["GRID"] = 7] = "GRID";
})(EMessageComponentType || (EMessageComponentType = {}));
export var EMessageSelectType;
(function (EMessageSelectType) {
    EMessageSelectType[EMessageSelectType["TEXT"] = 1] = "TEXT";
    EMessageSelectType[EMessageSelectType["USER"] = 2] = "USER";
    EMessageSelectType[EMessageSelectType["ROLE"] = 3] = "ROLE";
    EMessageSelectType[EMessageSelectType["CHANNEL"] = 4] = "CHANNEL";
})(EMessageSelectType || (EMessageSelectType = {}));
export var EMarkdownType;
(function (EMarkdownType) {
    EMarkdownType["TRIPLE"] = "t";
    EMarkdownType["SINGLE"] = "s";
    EMarkdownType["PRE"] = "pre";
    EMarkdownType["CODE"] = "c";
    EMarkdownType["BOLD"] = "b";
    EMarkdownType["LINK"] = "lk";
    EMarkdownType["VOICE_LINK"] = "vk";
    EMarkdownType["LINKYOUTUBE"] = "lk_yt";
})(EMarkdownType || (EMarkdownType = {}));
