// mezon-sdk/dist/esm/mezon-client-v2/structures/User.js
export class User {
    id;
    username;
    clan_nick;
    clan_avatar;
    display_name;
    avartar;
    dmChannelId;
    clan;
    channelManager;
    constructor(initUserData, clan, channelManager) {
        this.id = initUserData.id;
        this.avartar = initUserData.avartar;
        this.dmChannelId = initUserData?.dmChannelId;
        this.username = initUserData?.username;
        this.clan_nick = initUserData?.clan_nick;
        this.clan_avatar = initUserData?.clan_avatar;
        this.display_name = initUserData?.display_name;
        this.clan = clan;
        this.channelManager = channelManager;
    }
    async sendToken(sendTokenData) {
        const dataSendToken = {
            receiver_id: this.id,
            amount: sendTokenData.amount,
            note: sendTokenData?.note ?? "",
            extra_attribute: sendTokenData?.extra_attribute ?? "",
        };
        return this.clan.apiClient.sendToken(this.clan.sessionToken, dataSendToken);
    }
    async _createDmChannel() {
        const dmChannel = await this.channelManager.createDMchannel(this.id);
        return dmChannel ?? {};
    }
}
