// mezon-sdk/dist/esm/constants/enum.js
//Events socket will handle from MezonClient
// import { convertInternalEventToEvents } from "../utils/helper";
export var InternalEventsSocket;
(function (InternalEventsSocket) {
    InternalEventsSocket["VoiceStartedEvent"] = "voice_started_event";
    InternalEventsSocket["VoiceEndedEvent"] = "voice_ended_event";
    InternalEventsSocket["VoiceJoinedEvent"] = "voice_joined_event";
    InternalEventsSocket["VoiceLeavedEvent"] = "voice_leaved_event";
    InternalEventsSocket["ChannelCreatedEvent"] = "channel_created_event";
    InternalEventsSocket["ChannelDeletedEvent"] = "channel_deleted_event";
    InternalEventsSocket["ChannelUpdatedEvent"] = "channel_updated_event";
    InternalEventsSocket["ClanProfileUpdatedEvent"] = "clan_profile_updated_event";
    InternalEventsSocket["ClanUpdatedEvent"] = "clan_updated_event";
    InternalEventsSocket["StatusPresenceEvent"] = "status_presence_event";
    InternalEventsSocket["StreamPresenceEvent"] = "stream_presence_event";
    InternalEventsSocket["StreamData"] = "stream_data";
    InternalEventsSocket["ChannelMessage"] = "channel_message";
    InternalEventsSocket["MessageTypingEvent"] = "message_typing_event";
    InternalEventsSocket["MessageReactionEvent"] = "message_reaction_event";
    InternalEventsSocket["ChannelPresenceEvent"] = "channel_presence_event";
    InternalEventsSocket["LastPinMessageEvent"] = "last_pin_message_event";
    InternalEventsSocket["CustomStatusEvent"] = "custom_status_event";
    InternalEventsSocket["UserChannelAddedEvent"] = "user_channel_added_event";
    InternalEventsSocket["AddClanUserEvent"] = "add_clan_user_event";
    InternalEventsSocket["UserProfileUpdatedEvent"] = "user_profile_updated_event";
    InternalEventsSocket["UserChannelRemovedEvent"] = "user_channel_removed_event";
    InternalEventsSocket["UserClanRemovedEvent"] = "user_clan_removed_event";
    InternalEventsSocket["RoleEvent"] = "role_event";
    InternalEventsSocket["GiveCoffeeEvent"] = "give_coffee_event";
    InternalEventsSocket["RoleAssignEvent"] = "role_assign_event";
    InternalEventsSocket["TokenSend"] = "token_sent_event";
    InternalEventsSocket["ClanEventCreated"] = "clan_event_created";
    InternalEventsSocket["MessageButtonClicked"] = "message_button_clicked";
    InternalEventsSocket["StreamingJoinedEvent"] = "streaming_joined_event";
    InternalEventsSocket["StreamingLeavedEvent"] = "streaming_leaved_event";
    InternalEventsSocket["DropdownBoxSelected"] = "dropdown_box_selected";
    InternalEventsSocket["WebrtcSignalingFwd"] = "webrtc_signaling_fwd";
    InternalEventsSocket["Notifications"] = "notifications";
    InternalEventsSocket["QuickMenu"] = "quick_menu_event";
})(InternalEventsSocket || (InternalEventsSocket = {}));
export var Events;
(function (Events) {
    /** Listen to messages user sends on the  channel, thread */
    Events["ChannelMessage"] = "channel_message";
    /** Listen to user react to messages on the channel, thread */
    Events["MessageReaction"] = "message_reaction_event";
    /** Listen to user react to messages on the channel, thread */
    Events["UserChannelRemoved"] = "user_channel_removed_event";
    /** Listen to user leaved/removed in the channel */
    Events["UserClanRemoved"] = "user_clan_removed_event";
    /** Listen to user added in the channel */
    Events["UserChannelAdded"] = "user_channel_added_event";
    /** Listen to channel created */
    Events["ChannelCreated"] = "channel_created_event";
    /** Listen to channel deleted */
    Events["ChannelDeleted"] = "channel_deleted_event";
    /** Listen to channel updated */
    Events["ChannelUpdated"] = "channel_updated_event";
    /** Listen to clan create new role */
    Events["RoleEvent"] = "role_event";
    /** Listen to users give coffee to each other */
    Events["GiveCoffee"] = "give_coffee_event";
    /** Listen to assigning a role to user */
    Events["RoleAssign"] = "role_assign_event";
    /** Listen to user added in CLAN */
    Events["AddClanUser"] = "add_clan_user_event";
    /** Listen to user send token to each other */
    Events["TokenSend"] = "token_sent_event";
    /** Listen to clan create a new event */
    Events["ClanEventCreated"] = "clan_event_created";
    /** Listen to user lick a button on embed message */
    Events["MessageButtonClicked"] = "message_button_clicked";
    /** Listen to user joined a stream room */
    Events["StreamingJoinedEvent"] = "streaming_joined_event";
    /** Listen to user leaved a stream room */
    Events["StreamingLeavedEvent"] = "streaming_leaved_event";
    /** Listen to user selected a input dropdown*/
    Events["DropdownBoxSelected"] = "dropdown_box_selected";
    /** Listen to user accepted call 1-1 */
    Events["WebrtcSignalingFwd"] = "webrtc_signaling_fwd";
    /** Listen to start voice */
    Events["VoiceStartedEvent"] = "voice_started_event";
    /** Listen to end voice */
    Events["VoiceEndedEvent"] = "voice_ended_event";
    /** Listen to user join voice room */
    Events["VoiceJoinedEvent"] = "voice_joined_event";
    /** Listen to user leave voice room */
    Events["VoiceLeavedEvent"] = "voice_leaved_event";
    /** Listen to add friend */
    Events["Notifications"] = "notifications";
    /** Listen to add quick menu */
    Events["QuickMenu"] = "quick_menu_event";
})(Events || (Events = {}));
export var ChannelType;
(function (ChannelType) {
    ChannelType[ChannelType["CHANNEL_TYPE_CHANNEL"] = 1] = "CHANNEL_TYPE_CHANNEL";
    ChannelType[ChannelType["CHANNEL_TYPE_GROUP"] = 2] = "CHANNEL_TYPE_GROUP";
    ChannelType[ChannelType["CHANNEL_TYPE_DM"] = 3] = "CHANNEL_TYPE_DM";
    ChannelType[ChannelType["CHANNEL_TYPE_GMEET_VOICE"] = 4] = "CHANNEL_TYPE_GMEET_VOICE";
    ChannelType[ChannelType["CHANNEL_TYPE_FORUM"] = 5] = "CHANNEL_TYPE_FORUM";
    ChannelType[ChannelType["CHANNEL_TYPE_STREAMING"] = 6] = "CHANNEL_TYPE_STREAMING";
    ChannelType[ChannelType["CHANNEL_TYPE_THREAD"] = 7] = "CHANNEL_TYPE_THREAD";
    ChannelType[ChannelType["CHANNEL_TYPE_APP"] = 8] = "CHANNEL_TYPE_APP";
    ChannelType[ChannelType["CHANNEL_TYPE_ANNOUNCEMENT"] = 9] = "CHANNEL_TYPE_ANNOUNCEMENT";
    ChannelType[ChannelType["CHANNEL_TYPE_MEZON_VOICE"] = 10] = "CHANNEL_TYPE_MEZON_VOICE";
})(ChannelType || (ChannelType = {}));
export var ChannelStreamMode;
(function (ChannelStreamMode) {
    ChannelStreamMode[ChannelStreamMode["STREAM_MODE_CHANNEL"] = 2] = "STREAM_MODE_CHANNEL";
    ChannelStreamMode[ChannelStreamMode["STREAM_MODE_GROUP"] = 3] = "STREAM_MODE_GROUP";
    ChannelStreamMode[ChannelStreamMode["STREAM_MODE_DM"] = 4] = "STREAM_MODE_DM";
    ChannelStreamMode[ChannelStreamMode["STREAM_MODE_CLAN"] = 5] = "STREAM_MODE_CLAN";
    ChannelStreamMode[ChannelStreamMode["STREAM_MODE_THREAD"] = 6] = "STREAM_MODE_THREAD";
})(ChannelStreamMode || (ChannelStreamMode = {}));
export var TypeMessage;
(function (TypeMessage) {
    TypeMessage[TypeMessage["Chat"] = 0] = "Chat";
    TypeMessage[TypeMessage["ChatUpdate"] = 1] = "ChatUpdate";
    TypeMessage[TypeMessage["ChatRemove"] = 2] = "ChatRemove";
    TypeMessage[TypeMessage["Typing"] = 3] = "Typing";
    TypeMessage[TypeMessage["Indicator"] = 4] = "Indicator";
    TypeMessage[TypeMessage["Welcome"] = 5] = "Welcome";
    TypeMessage[TypeMessage["CreateThread"] = 6] = "CreateThread";
    TypeMessage[TypeMessage["CreatePin"] = 7] = "CreatePin";
    TypeMessage[TypeMessage["MessageBuzz"] = 8] = "MessageBuzz";
    TypeMessage[TypeMessage["Topic"] = 9] = "Topic";
    TypeMessage[TypeMessage["AuditLog"] = 10] = "AuditLog";
    TypeMessage[TypeMessage["SendToken"] = 11] = "SendToken";
    TypeMessage[TypeMessage["Ephemeral"] = 12] = "Ephemeral";
})(TypeMessage || (TypeMessage = {}));
